app/services/catalog/update_order_item.rb

```diff
@@ -46,12 +46,12 @@ def mark_item_based_on_status
       when "ok"
         case @payload["state"]
         when "completed"
-          @order_item.mark_completed("Order Item Complete", :external_url => fetch_external_url)
+          @order_item.mark_completed("Order Item Complete", :external_url => fetch_external_url, :service_instance_ref => @service_instance_id)
         when "running"
           @order_item.update_message("info", "Order Item being processed with context: #{@payload["context"]}")
         end
       when "error"
-        @order_item.mark_failed("Order Item Failed", :external_url => fetch_external_url)
+        @order_item.mark_failed("Order Item Failed", :external_url => fetch_external_url, :service_instance_ref => @service_instance_id)
```

Comment:
It looks like there is a possibility that `fetch_external_url` will be called before `service_instance_ref` is saved on the `@order_item`. Unless I'm reading this wrong.

Reply:
Yes, that's correct. That's how it works currently, i.e. when a task update comes back with a `Completed` state and it can get the service_instance_id from the task context. In that case, it'll get stored after pulling back the external_url.

When the task update comes in with the "error" state and we can't pull back an external_url, it will persist the service_instance_ref still so we can get it when (or if) another task update comes in.

This new logic will only get called if the external_url is nil and we have a service_instance_ref to go off of, so it should _never_ get called without an id to do the lookup.

Reply:
Refactored this a bit in a pairing session.